### PR TITLE
Big surface syntax graph datastructure refactor

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
wow more removes than adds right now!

File: hydroflow_lang/src/graph/serde_graph.rs [hunk 1]
```diff
@@ -466,4 +469,50 @@ impl SerdeGraph {
...
+                })
+                .unwrap_or(sg_nodes.len());
+
+            for (idx, node_id) in sg_nodes.iter().copied().enumerate() {
+                let is_pull = idx < pull_to_push_idx;
+                node_color_map.insert(node_id, if is_pull { Color::Pull } else { Color::Push });
+            }
+        }
+
+        node_color_map
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/serde_graph.rs:516]: this repeated code will be unified when SerdeGraph is subsumed

**Comment by MingweiSamuel** [comment]
Docs for unified graph datastructure; https://www.mingweisamuel.com/hydroflow/doc/hydroflow_lang/graph/struct.HydroflowGraph.html

File: hydroflow_lang/src/graph/flat_to_partitioned.rs [hunk 2]
```diff
@@ -121,51 +127,34 @@ fn make_subgraph_collect(
...
-    SlotMap<GraphSubgraphId, Vec<GraphNodeId>>,
 ) {
     // Algorithm:
     // 1. Each node begins as its own subgraph.
     // 2. Collect edges. (Future optimization: sort so edges which should not be split across a handoff come first).
     // 3. For each edge, try to join `(to, from)` into the same subgraph.
 
-    graph.assert_valid();
+    // TODO(mingwei):
+    // self.partitioned_graph.assert_valid();
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/flat_to_partitioned.rs:146]: hmm

**Comment by MingweiSamuel** [comment]
There was a bug in how serdegraph figured out what "colors" (push vs pull) operators are, which was fixed when I copied some of that algorithm out of `flat_to_partitioned` and into `SerdeGraph` itself

File: hydroflow_lang/src/graph/hydroflow_graph.rs [hunk 3]
```diff
@@ -0,0 +1,847 @@
...
+                    code = operator
+                        .to_token_stream()
+                        .to_string()
+                        .replace('&', "&amp;")
+                        .replace('<', "&lt;")
+                        .replace('>', "&gt;")
+                        .replace('"', "&quot;")
+                        .replace('\n', "<br>"),
+                ),
+                Node::Handoff { .. } => writeln!(write, r#"    {:?}{{"handoff"}}"#, key.data()),
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/hydroflow_graph.rs:838]: HANDOFF_NODE_STR

**Review by zzlk — APPROVED** [review]

**Pull request merged**